tbot/error.py

```diff
@@ -26,10 +26,37 @@ class TbotException(Exception):
     Base-class for all exceptions which are specific to tbot.
     """
 
-    pass
+
+class ApiViolationError(TbotException, RuntimeError):
+    """
+    Base-class for all exceptions that are raised due to wrong use of tbot's API.
+
+    Seeing this exception means that the calling code has a bug and needs to be
+    fixed. This is in contrast to :py:class:`tbot.error.Failure` which is the
+    base-class for exceptions raised due to unexpected occurrences during a
+    test-run.
+
+    The subclasses of ``ApiViolationError`` should detail in their
+    documentation how to fix them.
+
+    .. versionadded:: UNRELEASED
+    """
```

Comment:
Another point which irks me: A lot of code is currently raising things like `TypeError`, `ValueError`, `IndexError`, etc. when called with invalid arguments.  Intuitively, these should also also be a part of `ApiViolationError` but I'm not sure how to facilitate this best...

Just keep it as is, so these being separate?  Or creating a custom version of each one like

```python
class TbotValueError(ApiViolationError, ValueError):
  pass
```

The latter doesn't seem very pythonic to me...  But not having the exceptions subclassed from `ApiViolationError` means the grouping won't be that useful in practice, I think...

Reply:
`TypeError` and `ValueError` are expected when passing arguments of the wrong type or value to a function. I don't think you need to do anything special there since it's generally understood and it's documented [here](https://docs.python.org/3/library/exceptions.html#TypeError) and [here](https://docs.python.org/3/library/exceptions.html#ValueError).

For `IndexError`, I think that should only be raised if the user specifies an index, but for cases like the one below, I think a more specific error makes sense.
https://github.com/Rahix/tbot/blob/74dd21f335ad66b647082fa6b03f13ddb825af48/tbot/context.py#L217-L228

Reply:
Fair enough, this sounds reasonable.

Reply:
Added a `MachineNotFoundError` for this specific case of `IndexError`.